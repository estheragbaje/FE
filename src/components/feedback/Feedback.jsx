import React from 'react';
import { Query, withApollo } from 'react-apollo';
import propTypes from 'prop-types';
import { toast } from 'react-toastify';
import { GET_SURVEY_DETAILS } from '../../graphql/queries';
import { SAVE_FEEDBACK } from '../../graphql/mutations';
import { Container, Title, LoadIngIcon } from './styles';
import TextResponse from './TextResponse';
import RatingResponse from './RatingResposne/RatingResponse';

class Feedback extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      surveyId: '',
      answers: [],
      isLoading: false,
    };
  }

  componentDidMount() {
    const {
      match: {
        params: { surveyId },
      },
    } = this.props;
    if (surveyId) {
      this.setState({ surveyId });
    }
  }

  changeHandler = (event, questionId) => {
    const {
      target: { value },
    } = event;
    this.updateState(value, questionId);
  };

  ratingHandler = (value, questionId) => {
    this.updateState(value, questionId, true);
  };

  updateState = (value, questionId, rating = false) => {
    const { answers } = this.state;
    const answer = answers.find((ans) => ans.questionId === questionId);
    const updatedAnswer = {
      ...answer,
      ...(rating ? { rating: value.toString() } : { comment: value }),
    };
    const updatedAnswers = answers.map((currentAns) => {
      if (currentAns.questionId === questionId) {
        return updatedAnswer;
      }
      return currentAns;
    });
    this.setState({ answers: updatedAnswers });
  };

  initializeAnswers = (data) => {
    if (data) {
      const {
        getSurveyDetails: { questions },
      } = data;
      const { surveyId } = this.state;
      const answers = questions.map((q) => {
        if (q.type === 'rating') {
          return { questionId: q.id, surveyId, rating: '' };
        }
        return { questionId: q.id, surveyId, comment: '' };
      });
      this.setState({ answers });
    }
  };

  validateInput = () => {
    let isValid = 0;
    const { answers } = this.state;
    answers.forEach((ans) => {
      if ({}.hasOwnProperty.call(ans, 'rating') && ans.rating) {
        isValid += 1;
      } else if ({}.hasOwnProperty.call(ans, 'comment') && ans.comment) {
        isValid += 1;
      }
    });
    return !!isValid;
  };

  submitHandler = async (e) => {
    e.preventDefault();

    const { client } = this.props;
    const { answers } = this.state;
    const isValid = this.validateInput();
    if (isValid) {
      try {
        this.setState({ isLoading: true });
        const { data } = await client.mutate({
          mutation: SAVE_FEEDBACK,
          variables: { input: { responses: answers } },
        });
        if (data) {
          toast.success('Feedback sent, Thank you üòç');
          this.setState({ isLoading: false, answers: [] });
        }
      } catch (error) {
        toast.error('An error occurred trying to save your response');
        this.setState({ isLoading: false });
      }
    } else {
      toast.error('Kindly respond to the questions before submitting üôèüèª');
    }
  };

  render() {
    const { surveyId, isLoading } = this.state;
    return (
      <Container>
        <Query
          query={GET_SURVEY_DETAILS}
          variables={{ surveyId }}
          onCompleted={(data) => this.initializeAnswers(data)}
        >
          {({ data, loading, error }) => {
            if (loading) return <p>Loading....</p>;
            if (error) return <div>{error.toString()}</div>;
            const {
              getSurveyDetails: { title, questions },
            } = data;
            return (
              <>
                <Title>{title}</Title>
                {questions.map((q, i) => {
                  if (q.type === 'rating') {
                    return (
                      <div className="form-group" key={q.id}>
                        <RatingResponse
                          question={q}
                          handleRating={this.ratingHandler}
                          index={i}
                        />
                      </div>
                    );
                  }
                  return (
                    <div className="form-group" key={q.id}>
                      <TextResponse
                        question={q}
                        changeHandler={this.changeHandler}
                        index={i}
                      />
                    </div>
                  );
                })}
                <div className="form-group">
                  <button
                    className="btn btn-info btn-block mt-4"
                    type="submit"
                    onClick={this.submitHandler}
                  >
                    { isLoading ? 'processing... ' : 'Submit' }
                    {isLoading && <LoadIngIcon />}
                  </button>
                </div>
              </>
            );
          }}
        </Query>
      </Container>
    );
  }
}

Feedback.propTypes = {
  match: propTypes.shape({
    params: propTypes.shape({
      surveyId: propTypes.string.isRequired,
    }),
  }).isRequired,
  client: propTypes.shape({
    mutate: propTypes.func.isRequired,
  }).isRequired,
};

export default withApollo(Feedback);
