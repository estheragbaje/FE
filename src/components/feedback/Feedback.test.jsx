import React from 'react';
import { shallow } from 'enzyme';

import { Feedback } from './Feedback';

const mockProps = {
  match: {
    params: { surveyId: 'c71ca084-6c21-4ef8-b4fe-566ebb06c994' },
  },
  history: { push: jest.fn() },
  client: {
    query: jest.fn(),
    mutate: jest.fn(),
  },
};

const mockSurvey = {
  title: 'test title',
  id: '234455-2345',
  questions: [
    {
      id: 1,
      question: 'test rating question',
      type: 'rating',
    },
    {
      id: 2,
      question: 'test text question',
      type: 'text',
    },
  ],
};
const mockAnswers = [
  {
    questionId: 1,
    rating: '',
  },
  {
    questionId: 2,
    comment: '',
  },
];
let wrapper;
beforeAll(() => {
  // eslint-disable-next-line react/jsx-props-no-spreading
  wrapper = shallow(<Feedback {...mockProps} />);
});
describe('<Feedback />', () => {
  it('renders the Feedback component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should mount the component', () => {
    const instance = wrapper.instance();
    jest.spyOn(instance, 'fetchSurvey');
    jest.spyOn(instance, 'initializeAnswers');
    instance.componentDidMount();
    expect(instance.fetchSurvey).toHaveBeenCalledWith(
      mockProps.match.params.surveyId,
    );
    expect(instance.state.error).toBeDefined();
    expect(instance.initializeAnswers).not.toBeCalled();
  });
  it('should fetch initialize state if surveyId params is defined', () => {
    const instance = wrapper.instance();

    jest.spyOn(instance, 'fetchSurvey').mockImplementation(
      jest.fn(() => {
        instance.setState({ isLoading: false, survey: mockSurvey, answer: mockAnswers });
      }),
    );
    instance.componentDidMount();
    expect(instance.state.survey).toEqual(mockSurvey);
  });

  it('should not fetch survey if surveyId params is not defined', () => {
    // eslint-disable-next-line react/jsx-props-no-spreading
    const component = shallow(<Feedback {...{ match: { params: {} } }} />);
    const instance = component.instance();
    jest.spyOn(instance, 'fetchSurvey').mockImplementation(jest.fn());
    instance.componentDidMount();
    expect(instance.fetchSurvey).not.toBeCalled();
  });

  it('should render error if error occurs during survey fetch', () => {
    const instance = wrapper.instance();
    jest.spyOn(instance, 'fetchSurvey');
    instance.componentDidMount();
    expect(instance.state.error).toBeDefined();
  });

  it('should simulate changeHandler', () => {
    const event = { target: { value: 'dummy value' } };
    const instance = wrapper.instance();
    instance.setState({ answers: mockAnswers });
    jest.spyOn(instance, 'updateState');
    instance.changeHandler(event, 2);
    expect(instance.updateState).toHaveBeenCalledWith('dummy value', 2);
    expect(instance.state.answers[1]).toHaveProperty('comment', 'dummy value');
  });

  it('should simulate ratingHandler', () => {
    const instance = wrapper.instance();
    instance.setState({ answers: mockAnswers });
    jest.spyOn(instance, 'updateState');
    instance.ratingHandler('2', 1);
    expect(instance.updateState).toHaveBeenCalledWith('2', 1, true);
    expect(instance.state.answers[0]).toHaveProperty('rating', '2');
  });
});
