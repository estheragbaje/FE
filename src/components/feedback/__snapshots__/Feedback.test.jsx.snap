// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Feedback /> renders the Feedback component correctly 1`] = `
<styled.div>
  <Spinner />
  <p>
    Oops! Something went wrong
    <span
      aria-label="sad emoji"
      role="img"
    >
      ðŸ˜”
    </span>
  </p>
</styled.div>
`;
